internal/middleware/memoryprotection/memoryprotection.go

```diff
@@ -0,0 +1,271 @@
+	am.startBackgroundSampling()
+
+	log.Info().
+		Str("name", name).
+		Int64("memory_limit_bytes", memoryLimit).
+		Int("threshold_percent", config.ThresholdPercent).
+		Int("sample_interval_seconds", config.SampleIntervalSeconds).
+		Msg("memory protection middleware initialized with background sampling")
+
+	return am
+}
+
+// UnaryServerInterceptor returns a unary server interceptor that rejects incoming requests is memory usage is too high
+func (am *MemoryAdmissionMiddleware) UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		if err := am.checkAdmission(); err != nil {
+			_ = am.recordRejection(info.FullMethod)
+			return nil, err
+		}
+
+		return handler(ctx, req)
+	}
+}
+
+// StreamServerInterceptor returns a stream server interceptor that rejects incoming requests is memory usage is too high
+func (am *MemoryAdmissionMiddleware) StreamServerInterceptor() grpc.StreamServerInterceptor {
+	return func(srv any, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if err := am.checkAdmission(); err != nil {
+			_ = am.recordRejection(info.FullMethod)
+			return err
+		}
+
+		wrapped := middleware.WrapServerStream(stream)
+		return handler(srv, wrapped)
+	}
+}
+
+// checkAdmission returns an error if the request should be denied because memory usage is too high.
+func (am *MemoryAdmissionMiddleware) checkAdmission() error {
+	if am.memoryLimit < 0 || am.config.ThresholdPercent <= 0 {
+		// Memory protection is disabled
+		return nil
+	}
+
+	memoryUsage := am.getLastMemorySampleInBytes()
+
+	usagePercent := float64(memoryUsage) / float64(am.memoryLimit) * 100
+
+	// Metrics gauge is updated in background sampling
+
+	if usagePercent > float64(am.config.ThresholdPercent) {
+		return status.Errorf(codes.ResourceExhausted,
+			"server is experiencing memory pressure (%.1f%% usage, threshold: %d%%)",
+			usagePercent, am.config.ThresholdPercent)
+	}
+
+	return nil
+}
+
+// startBackgroundSampling starts a background goroutine that samples memory usage periodically
+func (am *MemoryAdmissionMiddleware) startBackgroundSampling() {
+	interval := time.Duration(am.config.SampleIntervalSeconds) * time.Second
+	ticker := time.NewTicker(interval)
+
+	go func() {
+		defer ticker.Stop()
+		// TODO this code might start running before the logger is setup, therefore we have a data race with the logger object :/
+		//defer log.Debug().Msg("memory protection background sampling stopped")
+		//
+		//log.Debug().
+		//	Dur("interval", interval).
+		//	Msg("memory protection background sampling started")
+
+		for {
+			select {
+			case <-ticker.C:
+				am.sampleMemory()
+			case <-am.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// sampleMemory samples the current memory usage and updates the cached value
+func (am *MemoryAdmissionMiddleware) sampleMemory() {
+	defer func() {
```

Comment:
In which scenarios may this happen? Can the process continue to operate reliably? I don't think we typically recover panics in SpiceDB. Does it, in theory, mean that at some point we could stop sampling and have SpiceDB operate on a stale snapshot of memory?